### PR TITLE
Changed spawn probability calculation to use conditional probabilities.

### PR DESCRIPTION
**Pull request by bwhitchurch**
Use conditional probabilities rather than flat x%.  Fixes the case when user specifies settings that total to more than 1

### CODE REVIEW

File: control.lua [hunk 1]
```diff
@@ -2,19 +2,45 @@ require("smallRuins")
...
+		largeChance / sumChance * totalChance + mediumThreshold
+
+    spawnTable = {small = smallThreshold, medium = mediumThreshold, large = largeThreshold}
+end
+
+-- Spawn Table won't change as long as settings don't change, so we can compute
+-- once and forget about it.
+script.on_init(spawnChances)
+script.on_load(spawnChances)
+script.on_event(defines.events.on_runtime_mod_setting_changed, spawnChances)
```
Comment (Bilka2) [inline control.lua:35]: Why did you choose to put it in a local variable, potentially producing desyncs, when it could be in `global.` instead?

Comment (bwhitchurch) [reply control.lua:35]: Are desyncs still a problem if it's using values that are map-wide settings? I can move it to the global table if it is an issue but I was also under the impression that local is preferred whenever possible.

Comment (bwhitchurch) [reply control.lua]: I think the latest commit should fix this.

**Comment by bwhitchurch** [comment]
I think I've now done this correctly. From what I understand in the documentation, we only ever need to change the values in the global table if mod settings change or we start a new map, so these are the times when we call `spawnChances()`. In any other situation all we want is a local reference to the global table which gets set up at the end of `spawnChances()` or during `on_load` if the game is being loaded without any settings changes. This should avoid any potential for desyncs since the data is only ever changed in the global copy, but also saves a tiny bit of performance by keeping a local reference, right?

**Comment by Bilka2** [comment]
I did needed cleanup and fixes on the branch and then merged this to master in #8. I detailed my reasoning for changes in the commit messages, so maybe give them a read. Thank you for contributing :)

**Comment by bwhitchurch** [comment]
Thanks, I think I understand the changes you made, and sorry about the formatting, I thought I had disabled my text editor's auto format. 
One question I have though is about the performance of `global`. I'm new to lua, but I've read in documentation that accessing true Global variables can take slightly longer than accessing local variables. Does this mean the `global` table for each mod is actually a local variable in the mod's environment or is it just that there are so few accesses of `spawn_table` that it doesn't really contribute to stack time that much? 
Thanks again!

**Comment by Bilka2** [comment]
> Does this mean the global table for each mod is actually a local variable in the mod's environment or is it just that there are so few accesses of spawn_table that it doesn't really contribute to stack time that much?

It's a global variable in the mod's environment. Another example of such a variable would be `math` (http://www.lua.org/manual/5.2/manual.html#6.6). Technically speaking, using a local reference to global.anything or math.anything is very very slightly faster, as you say.
However, I prefer to avoid on_load, where the local reference has to be set up. Furthermore, the code is not doing anything performance intensive, like calling some function every tick or whatever, so the very very small performance benefit really doesn't matter.

**Pull request merged**